lib/capybara/poltergeist/client/web_page.coffee

```diff
@@ -275,12 +275,23 @@ class Poltergeist.WebPage
       , selector
     )
 
+  getUserAgent: ->
+    this.native().settings.userAgent
+
   setUserAgent: (userAgent) ->
     this.native().settings.userAgent = userAgent
 
   getCustomHeaders: ->
     this.native().customHeaders
 
+  getPermanentCustomHeaders: ->
+    allHeaders = @getCustomHeaders()
```

Comment:
Should this be duplicating allHeaders before modifying it?

Reply:
The `customHeaders` getter actually returns a fresh copy each time ([here's the underlying C++ method](https://github.com/ariya/phantomjs/blob/master/src/webpage.h#L125), which returns a QVariantMap by value, not by reference).

If this weren't the case, then the spec "does not propagate temporary headers to new windows" would fail (because the temporary headers would get deleted when we call `@session.open_new_window`).

Reply:
Ah -- great